pkg/storage/snapshot/restore.go

```diff
@@ -596,6 +606,117 @@ func (t *vmRestoreTarget) reconcileDataVolumes() (bool, error) {
 	return createdDV || waitingDV, nil
 }
 
+func (t *vmRestoreTarget) getControllerRevision(namespace, name string) (*appsv1.ControllerRevision, error) {
+	revisionKey := cacheKeyFunc(namespace, name)
+	obj, exists, err := t.controller.CRInformer.GetStore().GetByKey(revisionKey)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, fmt.Errorf("Unable to find ControllerRevision %s", revisionKey)
+	}
+	return obj.(*appsv1.ControllerRevision), nil
+}
+
+func (t *vmRestoreTarget) getVirtualMachineSnapshot(namespace, name string) (*snapshotv1.VirtualMachineSnapshot, error) {
+	vmSnapshotKey := cacheKeyFunc(namespace, name)
+	obj, exists, err := t.controller.VMSnapshotInformer.GetStore().GetByKey(vmSnapshotKey)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, fmt.Errorf("Unable to find VirtualMachineSnapshot %s", vmSnapshotKey)
+	}
+	return obj.(*snapshotv1.VirtualMachineSnapshot), nil
+}
+
+func (t *vmRestoreTarget) restoreInstancetypeControllerRevision(VMSnapshotRevisionName, vmSnapshotName string, vm *kubevirtv1.VirtualMachine) (*appsv1.ControllerRevision, error) {
+	// Switch the snapshot and vm names for the restored CR name
+	restoredCRName := vm.Name + strings.Replace(VMSnapshotRevisionName, vmSnapshotName, "", 1)
+
+	// If the target VirtualMachine already exists it's likely that the original ControllerRevision is already present.
+	// Check that here by attempting to lookup the CR using the generated restoredCRName.
+	// Ignore any NotFound errors raised allowing the CR to be restored below.
+	if t.doesTargetVMExist() {
+		existingCR, err := t.getControllerRevision(vm.Namespace, restoredCRName)
+		if err != nil && !errors.IsNotFound(err) {
+			return nil, err
+		}
+		if existingCR != nil {
+			// TODO - Check the contents of the existing CR here against that of the snapshot CR
+			return existingCR, nil
+		}
+	}
```

Comment:
What about the case where `errors.IsNotFound(err) == true`? Shouldn't we log a warning in that case since it's not expected?

Reply:
I don't think so, while it's unexpected it's entirely possible for a user to mutate the original `VirtualMachine` in such a way that the original CR is no longer used and could even be deleted by the user. I'd rather not tbh.